First batch of fixes for issues found by pjdfstest.

marcinslusarz opened the pull request:
Each patch can be reviewed independently.

Fixes #207.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pmem/pmemfile/237)
<!-- Reviewable:end -->

GBuella commented on src/libpmemfile-posix/file.c:
```diff
@@ -359,6 +359,20 @@ _pmemfile_openat(PMEMfilepool *pfp, struct pmemfile_vinode *dir,
 		goto end;
 	}
 
+	if (flags & PMEMFILE_O_PATH)
```
So these dozen lines of code is moved from the end of the function to here. Why, by that way?

marcinslusarz replied:
In the next commit ("posix: clear SUID & SGID when writing to a file owned by someone else", which GitHub shows as earlier one for some reason) I need access to file->flags at earlier moment. I'll move this chunk of code to the relevant commit and fix ordering.

marcinslusarz replied:
Done.

GBuella replied:
I see now.

GBuella commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 11 of 11 files at r1, 1 of 1 files at r2, 7 of 7 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/237#-:-KoH2svfAcQ3xU4FxPYA:bnfp4nl)*
<!-- Sent from Reviewable.io -->

sarahjelinek commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[src/libpmemfile/CMakeLists.txt, line 52 at r3](https://reviewable.io:443/reviews/pmem/pmemfile/237#-KohGUwLSMT-A1fcpHa6:-KohGUwLSMT-A1fcpHa7:b3820kt) ([raw file](https://github.com/pmem/pmemfile/blob/d88352f09b9e3a94bc95f0b127129f7746ffdef6/src/libpmemfile/CMakeLists.txt#L52)):*
> ```CMake
> 		"libcap not found - needed by libpmemfile
> 		to skip building libpmemfile, set the BUILD_LIBPMEMFILE option to OFF")
> endif()
> ```

If libcap isn't found does that really mean we should fail to build libpmemfile?  Is this a requirement rather than a recommendation?

---

*[src/libpmemfile/preload.c, line 169 at r3](https://reviewable.io:443/reviews/pmem/pmemfile/237#-KohH8BLwmbDBbmLaIBy:-KohH8BLwmbDBbmLaIBz:bd98fxg) ([raw file](https://github.com/pmem/pmemfile/blob/d88352f09b9e3a94bc95f0b127129f7746ffdef6/src/libpmemfile/preload.c#L169)):*
> ```C
> 		fd = syscall_no_intercept(SYS_memfd_create, path, 0);
> 		/* memfd_create can fail for too long name */
> 		if (fd < 0) {
> ```

I am confused :-). If the name is too long here the name is too long in general for operating for a pmemfile_open(), correct? The path in this call is the path for the pmemfile?

---

*[src/libpmemfile/preload.c, line 1617 at r3](https://reviewable.io:443/reviews/pmem/pmemfile/237#-KohHZ1jt1VoRVyayUUq:-KohHZ1jt1VoRVyayUUr:b47a1aj) ([raw file](https://github.com/pmem/pmemfile/blob/d88352f09b9e3a94bc95f0b127129f7746ffdef6/src/libpmemfile/preload.c#L1617)):*
> ```C
> 		FATAL("!cap_get_proc");
> 
> 	cap_flag_value_t v;
> ```

'v' is not a great name for this variable. How about cap_val or something like that? In general one letter variable names are not descriptive enough.

---

*[src/libpmemfile-posix/file.c, line 439 at r3](https://reviewable.io:443/reviews/pmem/pmemfile/237#-KohJ78AmC6f2tOPCT7K:-KohJ78AmC6f2tOPCT7L:be1sogd) ([raw file](https://github.com/pmem/pmemfile/blob/d88352f09b9e3a94bc95f0b127129f7746ffdef6/src/libpmemfile-posix/file.c#L439)):*
> ```C
> 			os_rwlock_wrlock(&vinode->rwlock);
> 
> 			error = vinode_truncate(pfp, vinode, 0);
> ```

Why don't we check error here rather than after the if (file->flags & PFILE_WRITE) ?  Does an error here affect the next conditional block?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/237)*
<!-- Sent from Reviewable.io -->

marcinslusarz commented:
Review status: 13 of 19 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/libpmemfile/CMakeLists.txt, line 52 at r3](https://reviewable.io:443/reviews/pmem/pmemfile/237#-KohGUwLSMT-A1fcpHa6:-KolHBPtNSlleuzseLIe:bflh5l1) ([raw file](https://github.com/pmem/pmemfile/blob/d88352f09b9e3a94bc95f0b127129f7746ffdef6/src/libpmemfile/CMakeLists.txt#L52)):*
<details><summary><i>Previously, sarahjelinek (Sarah Jelinek) wrote…</i></summary><blockquote>

If libcap isn't found does that really mean we should fail to build libpmemfile?  Is this a requirement rather than a recommendation?
</blockquote></details>

Without libcap capability.h is not available, which means "update_capabilities" would generate compile errors.

---

*[src/libpmemfile/preload.c, line 169 at r3](https://reviewable.io:443/reviews/pmem/pmemfile/237#-KohH8BLwmbDBbmLaIBy:-KolHVO5kiyMrdI_nR1s:b-mh1ejh) ([raw file](https://github.com/pmem/pmemfile/blob/d88352f09b9e3a94bc95f0b127129f7746ffdef6/src/libpmemfile/preload.c#L169)):*
<details><summary><i>Previously, sarahjelinek (Sarah Jelinek) wrote…</i></summary><blockquote>

I am confused :-). If the name is too long here the name is too long in general for operating for a pmemfile_open(), correct? The path in this call is the path for the pmemfile?
</blockquote></details>

memfd_create has much smaller limit (than kernel/pmemfile-posix) on the number of characters (IIRC 128).

---

*[src/libpmemfile/preload.c, line 1617 at r3](https://reviewable.io:443/reviews/pmem/pmemfile/237#-KohHZ1jt1VoRVyayUUq:-KolLfa0TitNITXMClXh:bt2vr1b) ([raw file](https://github.com/pmem/pmemfile/blob/d88352f09b9e3a94bc95f0b127129f7746ffdef6/src/libpmemfile/preload.c#L1617)):*
<details><summary><i>Previously, sarahjelinek (Sarah Jelinek) wrote…</i></summary><blockquote>

'v' is not a great name for this variable. How about cap_val or something like that? In general one letter variable names are not descriptive enough.
</blockquote></details>

Fixed.

---

*[src/libpmemfile-posix/file.c, line 439 at r3](https://reviewable.io:443/reviews/pmem/pmemfile/237#-KohJ78AmC6f2tOPCT7K:-KolI9bztavjk-0bQJvg:b-hxc1d3) ([raw file](https://github.com/pmem/pmemfile/blob/d88352f09b9e3a94bc95f0b127129f7746ffdef6/src/libpmemfile-posix/file.c#L439)):*
<details><summary><i>Previously, sarahjelinek (Sarah Jelinek) wrote…</i></summary><blockquote>

Why don't we check error here rather than after the if (file->flags & PFILE_WRITE) ?  Does an error here affect the next conditional block?
</blockquote></details>

That's a bug. Fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/237)*
<!-- Sent from Reviewable.io -->

sarahjelinek commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 6 of 11 files at r1, 7 of 7 files at r3, 6 of 6 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/237#-:-KonfxmRYz0YVaJs6Abh:bnfp4nl)*
<!-- Sent from Reviewable.io -->
